lib/pulse/lib/Pulse.Lib.LinkedList.fst

```diff
@@ -462,66 +462,52 @@ let append_assoc_singleton (l1 l2:list 'a) (x:'a)
     [SMTPat (l1@(x::l2))]
 = List.Tot.Properties.append_assoc l1 [x] l2
 
-let trigger (x:'a) : slprop = emp
-
+fn not_is_last_cell (#t:Type) (x:llist t)
+    preserves is_list x 'l
+    requires pure (Cons? 'l)
+    returns b:bool
+    ensures pure (b == (List.Tot.length 'l <> 1))
+{
+  not (is_last_cell x)
+}
 
 fn append_iter (#t:Type) (x y:llist t)
   requires is_list x 'l1 ** is_list y 'l2 ** pure (Cons? 'l1)
   ensures is_list x ('l1 @ 'l2)
 {
-  some_iff_cons x;
   let mut cur = x;
   (* the base case, set up the initial invariant *)
   forall_intro_is_list_idem x;
   rewrite (forall* l. is_list x l @==> is_list x l)
       as  (forall* l. is_list x l @==> is_list x ([]@l));
-  while (
-    with _b ll pfx sfx. _;
-    let l = Pulse.Lib.Reference.(!cur);
-    some_iff_cons l;
-    let b = is_last_cell l;
-    if b 
-    { 
-      false
-    }
-    else 
-    {
-      let next = move_next_forall l;
-      with hd tl. _;
-      (* this is the key induction step *)
-      FA.trans_compose 
-          (is_list next) (is_list l) (is_list x)
-          (fun tl -> hd :: tl)
-          (fun tl -> pfx @ tl);
-      rewrite (forall* tl. is_list next tl @==> is_list x (pfx@(hd::tl)))
-           as (forall* tl. is_list next tl @==> is_list x ((pfx@[hd])@tl));
-      Pulse.Lib.Reference.(cur := next);
-      non_empty_list next; (* need to prove Some? next *)
-      true
-    }
-  )
-  invariant b.
-  exists* ll pfx sfx.
-      pts_to cur ll **
+  while (not_is_last_cell Pulse.Lib.Reference.(!cur))
```

Comment:
If I write `not (is_last_cell (!cur))` directly, then I get this error:
```
- Could not resolve all free variables in the proposition:
    _posth == op_Negation ?__anf2123
```

cc @nikswamy

Reply:
Minimized to #455.